Add the Skeleton type and its SkeletonManager 🦴

setzer22 opened the pull request:
<!-- Thank you for making a pull request! Below are the recommended steps to validate/complete your PR -->

## Checklist

- [x] cargo clippy reports no issues
- [x] cargo deny issues have been fixed or added to deny.toml
- [x] relevant examples/test cases run
- [x] changes added to changelog
  - [x] Add credit to yourself for each change: `Added new functionality @githubname`.

About the changelog, for now I only documented the breaking change in the API (the fact that `Object` now takes a `mesh_kind`). We can add another line on the changelog once the skinning system is finished :slightly_smiling_face: 

## Problems PR Solves
Continuing preparations for mesh skinning, this second PR adds a skeleton manager, and the `Skeleton` type.

Skeletons store the list of "joint deltas". This is the name I chose to refer to the transformation matrices that need to be applied to vertices, and it's not the same as the bone transformation matrices you'd usually get after running an animation system. To make this a bit more user friendly, there's two methods in `Renderer`: One sets the joint deltas directly, the other takes the global joint transforms plus the inverse bind matrices and multiplies them together. The method is not that useful once you understand mesh skinning, but I believe having it there is not a huge maintenance burden, and its docstring is good at educating users who don't know how to deal with this kind of mesh data.

Skeletons also own a range of vertices in the vertex buffer. That vertex range holds a complete duplicate of the original mesh this skeleton was created from, and will be used in the GPU skinning compute shader to modify the vertices in-place. The skeleton does not own an index buffer, instead, it can use the same index buffer as the underlying mesh.

Objects can now be created with a static mesh, or a skeleton (which has an underlying mesh). This is reflected by the change in `Object` which now has a `mesh_kind` field instead, which takes one of two variants: `Static` or `Animated`.



## Related Issues
#27

setzer22 committed:
Add initial version of the Skeleton Manager

setzer22 committed:
Add Skeleton type and start laying out the manager code

setzer22 committed:
Make sure joint data is copied during reallocation.

setzer22 committed:
Adapt examples

setzer22 committed:
Add instructions to create and modify skeleton data

setzer22 committed:
Cargo fmt

setzer22 committed:
Address clippy lints

setzer22 committed:
Update changelod. Add utility constructor in Skeleton

setzer22 commented on rend3/src/managers/mesh.rs:
```diff
@@ -229,6 +229,97 @@ impl MeshManager {
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_uv0,
+            original.vertex_range.start as u64,
+            &self.buffers.vertex_uv0,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_UV_SIZE) as u64,
+        );
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_uv1,
+            original.vertex_range.start as u64,
+            &self.buffers.vertex_uv1,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_UV_SIZE) as u64,
+        );
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_color,
+            original.vertex_range.start as u64,
+            &self.buffers.vertex_color,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_COLOR_SIZE) as u64,
+        );
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_joint_index,
+            original.vertex_range.start as u64,
+            &self.buffers.vertex_joint_index,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_JOINT_INDEX_SIZE) as u64,
+        );
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_joint_weight,
+            original.vertex_range.start as u64,
+            &self.buffers.vertex_joint_weight,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_JOINT_WEIGHT_SIZE) as u64,
+        );
```
@cwfitzgerald  Please double-check I didn't screw this up! :sweat_smile:

setzer22 commented on rend3/src/managers/mesh.rs:
```diff
@@ -335,6 +426,22 @@ impl MeshManager {
                 &new_vert_range,
                 VERTEX_COLOR_SIZE,
             );
+            copy_vert(
+                encoder,
+                &self.buffers.vertex_joint_index,
+                &new_buffers.vertex_joint_index,
+                mesh,
+                &new_vert_range,
+                VERTEX_JOINT_WEIGHT_SIZE,
+            );
+            copy_vert(
+                encoder,
+                &self.buffers.vertex_joint_weight,
+                &new_buffers.vertex_joint_weight,
+                mesh,
+                &new_vert_range,
+                VERTEX_JOINT_INDEX_SIZE,
+            );
```
This was missing from my previous PR.

setzer22 commented on rend3/src/managers/skeleton.rs:
```diff
@@ -0,0 +1,86 @@
+use super::MeshManager;
+
+/// Internal representation of a Skeleton
+#[repr(C, align(16))]
+#[derive(Debug, Clone)]
```
Not sure if this is needed here. I assume we're going to be sending some of this data to the compute shader, but probably not like this?

cwfitzgerald replied:
Yeah this doesn't need to be repr(C) as we can'tt send it to the gpu anyway as it's not copy

cwfitzgerald commented on rend3/src/managers/mesh.rs:
```diff
@@ -229,6 +229,97 @@ impl MeshManager {
+            &self.buffers.vertex_tangent,
+            vertex_range.start as u64,
+            (vertex_range.len() * VERTEX_TANGENT_SIZE) as u64,
+        );
```
~~We don't need to copy any of the ones the skinning shader is about to rewrite.~~

Oh we don't have a skinning shader yet... we should just flag this as to be removed.

setzer22 replied:
If we don't copy this data, when someone defines an animated mesh but forgets to add skinning to their pipeline they will get garbage. I can imagine this sort of thing happening when someone uses rend3_gltf (which will, at some point, autodetect animated meshes) but their custom rendering pipeline.

cwfitzgerald replied:
Fair enough.

cwfitzgerald commented on rend3/src/managers/mesh.rs:
```diff
@@ -229,6 +229,97 @@ impl MeshManager {
+
+        encoder.copy_buffer_to_buffer(
+            &self.buffers.vertex_position,
+            original.vertex_range.start as u64,
```
This isn't actually right, copy_buffer_to_buffer deals with bytes, and the allocator is allocating vertices.

setzer22 replied:
Oh, :man_facepalming:  I need to multiply by the size, right? It's so easy to mess up this stuff, thanks!

cwfitzgerald commented on rend3/src/managers/skeleton.rs:
```diff
@@ -0,0 +1,86 @@
+use rend3_types::{MeshHandle, RawSkeletonHandle, Skeleton, SkeletonHandle};
+use wgpu::{CommandEncoder, Device, Queue};
+
+use super::MeshManager;
```
Nit, combine this super into a `crate::` import with the registry import.

cwfitzgerald reviewed (changes requested):
Code in general looks quite good

setzer22 committed:
Use the right offsets in copy_buffer_to_buffer

setzer22 committed:
Internal skeleton doesn't need to be #[repr(C)]

setzer22 commented:
Alright, I think I took care of all the changes!

There's just a remaining comment about whether or not to copy those buffers. If we decide to remove them, I will do it on my next PR :+1:

cwfitzgerald commented on rend3/src/managers/mesh.rs:
```diff
@@ -229,6 +229,95 @@ impl MeshManager {
+            &self.buffers.vertex_position,
+            (original.vertex_range.start * VERTEX_POSITION_SIZE) as u64,
+            &self.buffers.vertex_position,
+            vertex_range.start as u64,
```
Have the same problem here :)

setzer22 replied:
:facepalm:  We're lucky that at least one of us is paying attention here... :sweat_smile: Thanks again

cwfitzgerald commented on rend3/src/renderer/mod.rs:
```diff
@@ -442,6 +457,46 @@ impl Renderer {
+    /// Note that this is not the same as the joint's transformation. See
+    /// [Renderer::set_skeleton_joint_positions] for an alternative method that
+    /// allows setting the joint transformation instead.
+    pub fn set_skeleton_joint_deltas(&self, handle: &SkeletonHandle, joint_deltas: Vec<Mat4>) {
```
Missing a #[track_caller]

cwfitzgerald commented on rend3/src/renderer/mod.rs:
```diff
@@ -442,6 +457,46 @@ impl Renderer {
+        joint_global_positions: &[Mat4],
+        inverse_bind_poses: &[Mat4],
+    ) {
+        let joint_deltas: Vec<Mat4> = joint_global_positions
```
It would be nice to factor both this and the other code in some way to get some code re-use. It's not a major deal, but I think would improve things a bit.

setzer22 replied:
I agree. I added an associated function to the `Skeleton` type which combines the matrices together and this is now used both in the Skeleton constructor and in the renderer

cwfitzgerald reviewed (changes requested):
One more round :)

setzer22 committed:
Address second round of review comments for #326

- Fix second copy_buffer_to_buffer offset being wrong.
- Refactor the logic computing the joint matrices into Skeleton
- Add missing #[track_caller]

cwfitzgerald commented:
Thanks this is shaping up quite nicely!
